nuxeo-routing-core/src/main/java/org/nuxeo/ecm/platform/routing/core/impl/DocumentRoutingServiceImpl.java

```diff
@@ -1034,6 +1038,41 @@ public DocumentModelList runQuery() {
         }
     }
 
+    /**
+     * Cancel the workflow instance if all its attached document don't exist anymore. If the workflow is cancelled then
+     * the isWowkflowCanceled is set to true.
+     *
+     * @since 8.4
+     */
+    class AttachedDocumentsChecker extends UnrestrictedSessionRunner {
+
+        String workflowInstanceId;
+
+        Boolean isWowkflowCanceled = false;
+
+        protected AttachedDocumentsChecker(CoreSession session, String workflowInstanceId) {
+            super(session);
+            this.workflowInstanceId = workflowInstanceId;
+        }
+
+        @Override
+        public void run() {
+            DocumentModel routeDoc = session.getDocument(new IdRef(workflowInstanceId));
+            DocumentRoute routeInstance = routeDoc.getAdapter(DocumentRoute.class);
+            if (routeInstance.getAttachedDocuments().isEmpty()) {
```

Comment:
Put `routeInstance.getAttachedDocuments()` in a local variable as it is used twice.